compiler-cli/src/dependencies.rs

```diff
@@ -543,6 +491,237 @@ fn same_requirements(
     Ok(left == right)
 }
 
+struct DependencyManager<Telem: Telemetry, P: dependency::PackageFetcher> {
+    runtime: tokio::runtime::Handle,
+    package_fetcher: P,
+    mode: Mode,
+    use_manifest: UseManifest,
+    telemetry: Telem,
+    check_major_versions: bool,
+}
+
+impl<Telem: Telemetry, P> DependencyManager<Telem, P>
+where
+    P: dependency::PackageFetcher,
+{
+    pub fn new(runtime: tokio::runtime::Handle, package_fetcher: P, telemetry: Telem) -> Self {
+        Self {
+            runtime,
+            package_fetcher,
+            mode: Mode::Dev,
+            use_manifest: UseManifest::No,
+            telemetry,
+            check_major_versions: false,
+        }
+    }
+
+    fn with_use_manifest(mut self, use_manifest: UseManifest) -> Self {
+        self.use_manifest = use_manifest;
+        self
+    }
+
+    fn with_mode(mut self, mode: Mode) -> Self {
+        self.mode = mode;
+        self
+    }
+
+    fn with_check_major_versions(mut self, check_major_versions: bool) -> Self {
+        self.check_major_versions = check_major_versions;
+        self
+    }
+
+    fn should_use_manifest(mut self) -> Self {
+        self.use_manifest = UseManifest::Yes;
+        self
+    }
```

Comment:
This is cool!

Another option which can be convenient at times is to have a struct with public fields and have an `into_dependency_manager()` method or something like that

Reply:
I looked into builder pattern and init pattern (misnamed but similar to what you're suggesting). But as the tokio runtime, Package Fetcher and Telemetry needs to be specified, I cannot implement Default. Which means I'll have to define all struct fields at callsite. Which means any addition or change to the struct will require changes at all callsites.

Alternatively I could go for a builder, but seems too small right now.

Reply:
Although...I could make it `DependencyManagerConfig.into_dependecy_manager(runtime: ..., package_fetcher: ..., telemetry: ...)`.